File: crabzilla-vavr-example1/crabzilla-vavr-example1-core/src/main/java/io/github/crabzilla/example1/customer/CustomerData.java
package io.github.crabzilla.example1.customer;

import io.github.crabzilla.core.DomainEvent;
import io.github.crabzilla.core.entity.EntityCommand;
import io.github.crabzilla.core.entity.EntityId;
import lombok.Value;

import java.time.Instant;
import java.util.UUID;

public class CustomerData {

  @Value
  public static class CustomerId implements EntityId {
    String stringValue;

    @Override
    public String stringValue() {
      return stringValue;
    }
  }

  // tag::commands[]

  @Value
  public static class CreateCustomer implements EntityCommand {
    UUID commandId;
    CustomerId targetId;
    String name;
  }

  @Value
  public static class ActivateCustomer implements EntityCommand {
    UUID commandId;
    CustomerId targetId;
    String reason;
  }

  @Value
  public static class CreateActivateCustomer implements EntityCommand {
    UUID commandId;
    CustomerId targetId;
    String name;
    String reason;
  }

  @Value
  public static class DeactivateCustomer implements EntityCommand {
    UUID commandId;
    CustomerId targetId;
    String reason;
  }

  // end::commands[]

  /**
   * This command is just for testing purposes
   */
  @Value
  public static class UnknownCommand implements EntityCommand {
    UUID commandId;
    CustomerId targetId;
  }

  // tag::events[]

  @Value
  public static class CustomerActivated implements DomainEvent {
    String reason;
    Instant when;
  }

  @Value
  public static class CustomerCreated implements DomainEvent {
    CustomerId id;
    String name;
  }

  @Value
  public static class CustomerDeactivated implements DomainEvent {
    String reason;
    Instant when;
  }

  // end::events[]

}

File: crabzilla-core/src/test/java/io/github/crabzilla/core/entity/example1/customer/CustomerData.kt
package io.github.crabzilla.example1.customer

import io.github.crabzilla.core.DomainEvent
import io.github.crabzilla.core.entity.EntityCommand
import io.github.crabzilla.core.entity.EntityId
import java.time.Instant
import java.util.*

data class CustomerId(val id: String) : EntityId {
  override fun stringValue(): String {
    return id
  }
}

// tag::events[]

data class CustomerCreated(val id: CustomerId, val name: String) : DomainEvent

data class CustomerActivated(val reason: String, val _when: Instant) : DomainEvent

data class CustomerDeactivated(val reason: String, val _when: Instant) : DomainEvent

// end::events[]

// tag::commands[]

data class CreateCustomer(override val commandId: UUID, override val targetId: CustomerId,
                          val name: String) : EntityCommand

data class ActivateCustomer(override val commandId: UUID, override val targetId: CustomerId,
                            val reason: String) : EntityCommand

data class DeactivateCustomer(override val commandId: UUID, override val targetId: CustomerId,
                              val reason: String) : EntityCommand

data class CreateActivateCustomer(override val commandId: UUID,
                                  override val targetId: CustomerId,
                                  val name: String, val reason: String) : EntityCommand

// end::commands[]

data class UnknownCommand(override val commandId: UUID, override val targetId: CustomerId)
  : EntityCommand
